### ISSUE TITLE
Building DDG from function invocation

### ISSUE THREAD

**Issue by margoseltzer** [End-to-end-provenance/RDataTracker#129]
It's possible that I am such a naive R user that this is a really stupid thing to want to do, but here is something I ran into and I wondered if others thought this would be a useful feature.

I was debugging some scripts and really wanted to create a DDG for a particular function invocation (i.e., I'm typing into the console something like foo(A, B, C)).  It seems like right now, the assumption is that you generate DDGs from scripts (and I could certainly create a script that did nothing more than invoke my function, but that doesn't seem particularly efficient).

So, I wondered - is this just something a "real" R user would never do or might this be something useful?

**Comment by erboose** [comment]
Console mode to the rescue!  If you want to capture a console session, just enter ddg.init() to initiate creation of a DDG and ddg.save() to save it.

There are some limitations, though. Because this uses the R history file, it's possible to loose information in long sessions. Also variable values are only retrieved when a call to RDT is made.  And some of the more advanced features of ddg.run, such as documentation of function internals, are not available here.

This feature has not been tested much recently. Let us know if it still works!

Sent from my iPhone

On Jul 8, 2016, at 8:23 AM, margoseltzer <notifications@github.com<mailto:notifications@github.com>> wrote:


It's possible that I am such a naive R user that this is a really stupid thing to want to do, but here is something I ran into and I wondered if others thought this would be a useful feature.

I was debugging some scripts and really wanted to create a DDG for a particular function invocation (i.e., I'm typing into the console something like foo(A, B, C)). It seems like right now, the assumption is that you generate DDGs from scripts (and I could certainly create a script that did nothing more than invoke my function, but that doesn't seem particularly efficient).

So, I wondered - is this just something a "real" R user would never do or might this be something useful?

-

**Comment by erboose** [comment]
Looks like console mode is no longer working after recent changes. I'll create an issue.

**Comment by margoseltzer** [comment]
> On Jul 8, 2016, at 8:55 AM, erboose <notifications@github.com> wrote:
> 
> Console mode to the rescue! If you want to capture a console session, just enter ddg.init() to initiate creation of a DDG and ddg.save() to save it.
> 
> There are some limitations, though. Because this uses the R history file, it's possible to loose information in long sessions. Also variable values are only retrieved when a call to RDT is made. And some of the more advanced features of ddg.run, such as documentation of function internals, are not available here.
> 
> This feature has not been tested much recently. Let us know if it still works!


I tried running this on an old version, and the inability to go into functions is the real show-stopper.  Here is my use case:

I have a function I would like to debug — either it’s giving me wrong output or producing an error.  So, I’d like to use the DDG to figure out why. Unfortunately, without going into the function, the graph I get back isn’t very helpful (it just says “you called this function”).

- M

**Comment by blernermhc** [comment]
Yes, that is a limitation of console mode.  We do not modify the interpreter at all and so our code does not run inside the REPL.  The ddg that is built in console mode is based on the history, which only records top-level commands.

**Comment by margoseltzer** [comment]
I’m not entirely sure we should close this one — while we can’t do it now, I think it’s a relatively important feature we might want to brainstorm around to see if we can figure out a solution

- M

**Comment by blernermhc** [comment]
We no longer use the history file in console mode, but using a tracing feature.  Still, that would only give us information when a top-level command is executed, not function internals.
